### ISSUE TITLE
Enable 512px tile sources

### ISSUE THREAD

**Issue by matteblair** [tangrams/tangram-es#1435]
**Background**

We currently assume that a tile from tiled data sources is intended to be viewed at a scale of 256 logical pixels per side. As the resolution of mobile devices increases, we incur a corresponding increase in the number of URL requests needed to cover a screen-sized area. A data source can reduce this number of requests by offering tiles meant for viewing at 512 logical pixels per side. These 512px tiles would cover the area of 4 256px tiles and would therefore reduce the number of URL requests needed for a typical view by a factor of 4. This would also result in fetching somewhat more off-screen data than with 256px tiles, but this trade-off is often beneficial in practice.

**Implementation**

The logical size of a tile is an attribute of the data source that provided that tile, and this size cannot be reliably inferred from the tile data itself. Therefore the tile size of a data source is probably best configured in the scene file within a `sources` entry. The syntax for this could be simply:
```yaml
sources:
  _myTileSource:
    tile_size: 512 # Defaults to 256 if absent.
```
Or, recognizing that sizes will almost certainly be powers of 2, we could specify an exponent relative to 256 instead:
```yaml
sources:
  _myTileSource:
    tile_scale: 2 # Defaults to 1 if absent, 2 denotes 512px, 3 denotes 1024px, etc.
    # Or maybe 0 denotes 256px, 1 denotes 512px.
```

Within the engine, we'll need additional logic for determining the display size of a tile based on the size specified in its data source configuration. The semantics of a Tile Coordinate will remain the same - that is, a Tile Coordinate {x, y, z} will denote the same geographic area for tiles of any pixel size. This means that the pixel size specified for a tile is effectively a measure of the _level of detail_ present in the tile data. Therefore, for a view covering the Tile Coordinate {x, y, z} using a data source of 512px tiles, we would display the tile {x/2, y/2, z-1} so that we show the same level of detail as we would for 256px tiles.

**Comment by nvkelso** [comment]
@bcamper Is there an issue tracking this on the JS side? I remember you have a branch for this, but don't see it pushed to the server?

**Comment by nvkelso** [comment]
The new (undocumented) endpoint for 512px tiles is:

- `tile.mapzen.com/{user}/{tileset}/v{version}/{tilesize}/{layer}/{z}/{x}/{y}.{format}`

For example:

- `tile.mapzen.com/mapzen/vector/v1/512/all/{z}/{x}/{y}.mvt`

**Comment by bcamper** [comment]
The JS branch is here: https://github.com/tangrams/tangram/compare/zoom-bias

I'm not sure there is an issue so I can create one parallel to this.

Since we will likely limit this to power-of-2 tile sizes for simplicity, I think either an exponent (as in the example) would work, or if we prefer to have the pixel scale specified explicitly, we could calculate the exponent scale, and only allow power-of-2 values.

**Comment by tallytalwar** [comment]
I vote for the exponent as tile sale as mentioned in the issue description.

**Comment by nvkelso** [comment]
I suspect most marketing around tile size is in pixels (like 256, 512,
1024), so prefer to express it in pixel size in the scene file config, with
Tangram figuring out the scale internally as a black box.

**Comment by tallytalwar** [comment]
Some quick initial work here for ES: https://github.com/tangrams/tangram-es/tree/512-tiles

**Comment by matteblair** [comment]
🏃 💨 ⏱  Speedy work!

I also prefer the more straightforward "tile size" syntax, it seems very hard to misinterpret. It would also make sense to do some input validation engine-side, since NPOT tile sizes might behave badly.

**Comment by bcamper** [comment]
So:

- `tile_size` must be greater than or equal to 256
- `tile_size`must be a power of 2
- Internally, the number of zoom levels to adjust by is:
  - `log2(tile_size) - log2(256)` (aka `log2(tile_size) - 8` :)

**Comment by bcamper** [comment]
added parsing for `tile_size` in JS, according to rules above, here: https://github.com/tangrams/tangram/commit/c09a4a8ebfa8d4726fe8b2f28342de7e4ffe3b0d

**Comment by bcamper** [comment]
Note I allowed `tile_size` to accept a number with or without units, e.g.

`tile_size: 512`
`tile_size: 512px`

are both acceptable

**Comment by tallytalwar** [comment]
@bcamper Thanks, can add that as well on ES.

**Comment by matteblair** [comment]
Hmm why accept suffixes? If we wanted to be really explicit with the units I'd say label the field `tile_size_px`, but I don't feel like we need to bother with units at all.

**Comment by bcamper** [comment]
I just think it's more consistent with other places we accept numeric
values? While it's the only units possible, it seems like it will lessen
confusion since other parameters that take pixels use `px` (and similarly,
sometimes with default units).

**Comment by matteblair** [comment]
There's a much broader discussion to have here about the challenges of specifying the scene file interface, but to put it simply: I will advocate for a _stricter contract_ on the behavior of the scene file because I believe it reliably results in more consistent and predictable behavior across implementations.

For example, https://github.com/tangrams/tangram/pull/548/commits/e6537ba57e0a45083dcb2ce31f465388ac4d26c5 will not only allow the inputs `512` and `512px`, but also `512 px`, `512PX`, `512m`, `512xp`, etc. Making sure that tangram-es behaves consistently with _wide contracts_ like this is challenging and, I believe, not the best use of our time.

**Comment by bcamper** [comment]
Fair point about the JS implementation accepting weird string values -- that was sloppy :)

However, I'm less convinced it makes sense to deviate from other places where we accept pixel-unit values. As far as I'm aware, those all accept either explicit units (`16px`) or have default units (which can be meters or pixels depending on the parameter). None of those include units in the parameter name like `tile_size_px` would. It seems like the more consistent but still relatively reasonable interface is to strictly accept `tile_size: 512` or `tile_size: 512px`, but nothing else.

On the JS side I should have used the existing unit parsing code for consistency. It is stricter about only accepting `px`, though I just noticed it does *trim* the value so it would accept `16 px` :/ I would be comfortable tightening that parsing in an upcoming release for simplicity (assuming we don't have this behavior in ES).

https://github.com/tangrams/tangram/blob/master/src/styles/style_parser.js#L212-L218

**Comment by matteblair** [comment]
Sprite definitions, for instance, allow only numeric values with no units, because there is only one rational unit to use. This seems like a similar case.

**Comment by matteblair** [comment]
If we were to define and follow a specific format for value-plus-optional-unit fields throughout the scene file, I would feel a lot better about using them!

**Comment by bcamper** [comment]
> If we were to define and follow a specific format for value-plus-optional-unit fields throughout the scene file, I would feel a lot better about using them!

What do you think is lacking in that regard now? e.g. which are the places where you see inconsistency, is that in response to my comment on JS allowing whitespace trimming (reasonable to remove) or other?

**Comment by matteblair** [comment]
I think we're actually _mostly_ following consistent rules for value-plus-optional-unit and I was just unclear about these rules because they're not explicitly stated anywhere. Here's what happens now when I use the following values for the `width` of a `lines` rule:

| Width | Tangram Play result | Tangram ES result |
|-------|---------------------|--------------------|
| 1 | 1 meter width | same |
| 1px | 1 pixel width | same |
| 1 px | 1 pixel width |  1 meter width |
| 1m | 1 meter width | same |
| 1 m | 1 meter width | same |
| 1z | 1 meter width | same |

I'm not sure whether these behaviors are all intentional, but practically speaking it seems like whitespace is the only inconsistency. I would prefer generating errors or at least warnings for invalid suffixes, but I think we've digressed enough in this issue already :)

Let's allow the `px` suffix for `tile_size` and document the rules governing unit suffixes.

**Comment by bcamper** [comment]
I dunno, your sprite example is a good one... I'm wavering :)

**Comment by lygstate** [comment]
Are that 1024px are possible?

**Comment by matteblair** [comment]
@lygstate Yes, this would allow 1024 pixel tiles in the same way.

**Comment by lygstate** [comment]
@matteblair I am thinking about generating 1024 pixel tiles in tilzen. So ther would be only 14 zoom level, and there is a possibility to pre-generating all the tiles into mbtiles.

**Comment by nvkelso** [comment]
@lygstate That should be possible in Tilezen by setting a larger `metatile` size, but we haven't tried it ourselves yet so there are likely to be some performance tuning &/or other gotchas we can help you thru.

**Comment by bcamper** [comment]
OK I flipped my position here :)

Let's make `tile_size` behave like `sprites` for now: only plain numbers without units accepted.

Then we can support optional `px` units with defaults on both as a separate follow-up.

**Comment by matteblair** [comment]
Released in 0.7.0
